qiskit/algorithms/amplitude_amplifiers/grover.py

```diff
@@ -156,31 +162,81 @@ def __init__(
         else:
             self._iterations = iterations
 
+        if quantum_instance is not None and sampler is not None:
+            raise ValueError("Only one of quantum_instance or sampler can be passed, not both!")
+
+        # check positionally passing the sampler in the place of quantum_instance
+        # which will be removed in future
+        if isinstance(quantum_instance, BaseSampler):
+            sampler = quantum_instance
+            quantum_instance = None
+
         self._quantum_instance = None
         if quantum_instance is not None:
-            self.quantum_instance = quantum_instance
+            warnings.warn(
+                "The quantum_instance argument has been superseded by the sampler argument. "
+                "This argument will be deprecated in a future release and subsequently "
+                "removed after that.",
+                category=PendingDeprecationWarning,
+                stacklevel=2,
+            )
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", category=PendingDeprecationWarning)
+                self.quantum_instance = quantum_instance
+
+        self._sampler = sampler
 
         self._sample_from_iterations = sample_from_iterations
         self._iterations_arg = iterations
 
     @property
+    @deprecate_function(
+        "The Grover.quantum_instance getter is pending deprecation. "
+        "This property will be deprecated in a future release and subsequently "
+        "removed after that.",
+        category=PendingDeprecationWarning,
+    )
     def quantum_instance(self) -> Optional[QuantumInstance]:
-        """Get the quantum instance.
+        """Pending deprecation: Get the quantum instance.
         Returns:
             The quantum instance used to run this algorithm.
         """
         return self._quantum_instance
 
     @quantum_instance.setter
+    @deprecate_function(
+        "The Grover.quantum_instance setter is pending deprecation. "
+        "This property will be deprecated in a future release and subsequently "
+        "removed after that.",
+        category=PendingDeprecationWarning,
+    )
     def quantum_instance(self, quantum_instance: Union[QuantumInstance, Backend]) -> None:
-        """Set quantum instance.
+        """Pending deprecation: Set quantum instance.
         Args:
             quantum_instance: The quantum instance used to run this algorithm.
         """
         if isinstance(quantum_instance, Backend):
             quantum_instance = QuantumInstance(quantum_instance)
         self._quantum_instance = quantum_instance
 
+    @property
+    def sampler(self) -> Optional[BaseSampler]:
+        """Get the sampler.
+
+        Returns:
+            The sampler used to run this algorithm.
+        """
+        return self._sampler
+
+    @sampler.setter
+    def sampler(self, sampler: Optional[BaseSampler]) -> None:
```

Comment:
The quantum_instance setter never included Optional so you could set a valid instance later but we never expected you to put it back it None contrary to the typehint - though of course nothing truly prevented that.

Reply:
right, since the quantum_instance setter does not include Optional, let's remove it. (The rare use case is that users might want to change a backend to a sampler from a quantum instance. To do that, we need to change the quantum_instance setter as well. I think it's not worth the effort. Users can make another `Grover` instance anytime).